### PR TITLE
feat(#31): implemented api for environment crud operations

### PR DESCRIPTION
**Pull request by akhilmhdh**
## What does it do

This will enable workspace environment crud operations

## Screenshots

<details>

<summary> Expand</summary>

## List Env
Screenshot 2023-01-11 at 11 06 19 PM

## Create Env
Screenshot 2023-01-11 at 11 06 24 PM

### Update Env
Screenshot 2023-01-11 at 11 06 47 PM

### Delete Env
Screenshot 2023-01-11 at 11 06 59 PM

### Dashboard
Screenshot 2023-01-11 at 11 07 16 PM

## Service Token
Screenshot 2023-01-11 at 11 07 40 PM

## Integration Page

Screenshot 2023-01-13 at 9 06 59 PM

Screenshot 2023-01-13 at 9 07 02 PM

</details>

### CODE REVIEW

**Comment by mv-turtle** [comment]
This is awesome! Makes total sense to me! 

I feel like @dangtony98 might say that the routes should be like 'workspace/:workspaceId/environment'. I know that he is doing some v2 revamp which should create a common way of naming routes. @dangtony98 do you have any opinions here?

File: backend/src/app.ts [hunk 1]
```diff
@@ -92,6 +93,7 @@ app.use('/api/v1/integration-auth', v1IntegrationAuthRouter);
 // v2 routes
 app.use('/api/v2/workspace', v2WorkspaceRouter);
 app.use('/api/v2/secret', v2SecretRouter);
+app.use('/api/v2/environment', v2EnvironmentRouter);
```
Comment (Zamion101) [inline backend/src/app.ts:96]: should be `/api/v2/environments`

File: backend/src/controllers/v2/environmentController.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    );
+    if (!workspace) {
+      throw new Error('Failed to update workspace');
+    }
+
+    await Secret.updateMany(
+      { workspace: workspaceId, environment: oldEnvironmentSlug },
+      { environment: environmentSlug }
+    );
+    await ServiceToken.updateMany(
```
Comment (Zamion101) [inline backend/src/controllers/v2/environmentController.ts:87]: Above and following database operations should follow `Atomicity` from ACID. This applies for similar operations further in the PR.

File: backend/src/routes/v2/environment.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    acceptedStatuses: [COMPLETED, GRANTED],
+  }),
+  param('workspaceId').exists().trim(),
+  body('environmentSlug').exists().trim(),
+  body('environmentName').exists().trim(),
+  validateRequest,
+  environmentController.createWorkspaceEnvironment
+);
+
+router.put(
```
Comment (Zamion101) [inline backend/src/routes/v2/environment.ts:26]: Correct HTTP Method should be `PATCH`.

"When a client needs to replace an existing Resource entirely, they can use PUT. When they're doing a partial update, they can use HTTP PATCH."

We're renaming just **one** environment variable and not entire list of environments.

**Review by Zamion101 — CHANGES_REQUESTED** [review]
There needs some changes and further discussion on how to handle envrionments in upcoming V2 API by @dangtony98

File: backend/src/controllers/v2/environmentController.ts (hunk 2 above)
Comment (akhilmhdh) [reply backend/src/controllers/v2/environmentController.ts:87]: Can you be more elaborate your point. All the above operations are atomic by nature from Mongodb 4.0 onwards.

File: backend/src/routes/v2/environment.ts (hunk 3 above)
Comment (akhilmhdh) [reply backend/src/routes/v2/environment.ts:26]: Not really. GIven the route scope is environments and not list of environments. 

An environment here contains slug and name. Patch would have asked user to update either slug or name, but that's not intentded here.  We are asking user to provide both the parameters together that is all entity. That is PUT not patch.

**Comment by akhilmhdh** [comment]
> There needs some changes and further discussion on how to handle envrionments in upcoming V2 API by @dangtony98

@Zamion101 And yes. I have talked with tony and now this PR is in a waiting for v2 routes to come inti picture and launched

File: backend/src/controllers/v2/environmentController.ts (hunk 2 above)
Comment (akhilmhdh) [reply backend/src/controllers/v2/environmentController.ts:87]: @Zamion101 Ohhh sure that's valid issue. But the catch is for mongodb to support transactions, its not simple as sql db. U need to enable replication to support transactions. For that we need to change the image's doc and also prod i guess.

I think then the scope gets bigger for this PR. Would love to get input from @dangtony98 @maidul98 on this.

Comment (Zamion101) [reply backend/src/controllers/v2/environmentController.ts:87]: I don't think we will have any problem as it is right now. I checked the dependencies and package versions and we're good to use following method from Mongoose documents;

https://mongoosejs.com/docs/transactions.html

Comment (akhilmhdh) [reply backend/src/controllers/v2/environmentController.ts:87]: @Zamion101 Yah library wise is ok all good.

But to support transactions in mongodb there is a architecture change in db required.

Check this: https://www.mongodb.com/community/forums/t/why-replica-set-is-mandatory-for-transactions-in-mongodb/9533

**Comment by mv-turtle** [comment]
Looking into this right now! Thank you for such detailed description!

File: frontend/pages/api/workspace/getAWorkspace.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  name: string;
+  organization: string;
+  environments: Array<{ name: string; slug: string }>;
+}
+
+/**
+ * This route lets us get the workspaces of a certain user
+ * @returns
+ */
+const getAWorkspace = (workspaceID:string) => {
```
Comment (mv-turtle) [inline frontend/pages/api/workspace/getAWorkspace.ts:15]: I think this is a duplicate of GetProjectInfo.js
We kinda use workspaces and projects interchangeably which id pretty confusing

**Review by mv-turtle — APPROVED** [review]

**Pull request merged**